site/drafts/withered.md

```diff
@@ -0,0 +1,431 @@
+---
+title: 'Composable filters using Witherable optics'
+author: "Chris Penner"
+date: "Oct 27, 2020"
+tags: [haskell, optics]
+description: ""
+---
+
+One of my favourite things about Haskell is that its structures and abstractions
+are very principled, and they have laws dictating correct behaviour.
+
+In my experience, this means that when you find a new way to piece together those
+abstractions it _almost always_ ends up doing something reasonable... or at the very least interesting!
+
+It comes as no surprise to some of you that I think about **optics** (lenses) an awful lot! 
+Optics are the most beautiful, composable, expressive, and adaptable abstractions I've
+ever had the pleasure of discovering. Optics provide an elegant abstraction
+for many data-manipulation strategies that have been around for decades.
+
+In this post we're going to explore a new type of optic which provides composable
+and expressive means for filtering complex data structures.
+To do so is going to require a base-level understanding of optics, and preferably readers would
+understand what a Traversal is. Bonus points if you've implemented one on your own or have some idea of how they work!
+
+## The Background
+
+Let's take a look at the Traversal typeclass, it's where we find Haskell's
+all-powerful secret weapon **traverse**! ([BTW The answer is always traverse.](https://impurepics.com/fp-bot/index.html))
+
+```haskell
+class (Functor t, Foldable t) => Traversable t where
+  traverse :: Applicative f => (a -> f b) -> t a -> f (t b)
+```
+
+Many beginners rightfully stumble on this function because it's just so dang abstract! It's tough to tell what it does for you, and how you could use it (the answer is almost anywhere).
+
+
+Eventually, folks realized that we could write more "specialized" versions of this type which worked on anything, not just Traversable datatypes, and the Traversal was born!
+
+A traversal is anything that fits this rough shape:
+
+```haskell
+forall f. Applicative f => (a -> f b) -> (s -> f t)
```

Comment:
One way i like explaining this datatype, and it is worth explaining how you got here, is that we're "squashing" information in some sense about the input and output data into a new parameter. I only mention this because it's actually rare to see people explain this step, and because the mnemonic "s = source and t = target" isn't very clear unless someone says it! In this case, the source and target lose the ability to talk about what's inside the traversable object, only dealing with its external traversable structure. *That* is why it's "specialized".